### PR TITLE
[CGP-438] Log text messages in WalletAPI

### PR DESCRIPTION
**Pull request by j-mueller**
* Add `LoggingWallet` type with the ability to log messages and throw
  errors

### CODE REVIEW

File: wallet-api/src/Wallet/API.hs [hunk 1]
```diff
@@ -195,6 +201,22 @@ data WalletAPIError =
...
+    deriving (Eq, Ord, Show, Generic, Semigroup, Monoid)
+
+instance FromJSON WalletLog
+instance ToJSON WalletLog
+
+-- | Write some information to the log
+logMsg :: MonadWriter WalletLog m => Text -> m ()
+logMsg = tell . WalletLog . pure
+
+-- | WalletAPI with the ability to log messages and throw errors
```
Comment (michaelpj) [inline wallet-api/src/Wallet/API.hs]: wrong type

File: wallet-api/src/Wallet/API.hs [hunk 2]
```diff
@@ -195,6 +201,22 @@ data WalletAPIError =
...
+type LoggingWallet m = (
```
Comment (michaelpj) [inline wallet-api/src/Wallet/API.hs]: I don't like the name given that it includes errors too. How about `WalletDiagnostics`? I'd also probably not imply `WalletApi` for clarity.

Comment (michaelpj) [reply wallet-api/src/Wallet/API.hs]: Although that's a bit subjective.

File: wallet-api/src/Wallet/Emulator/Types.hs [hunk 3]
```diff
@@ -93,8 +95,25 @@ data Notification = BlockValidated Block
                   deriving (Show, Eq, Ord)
 
 -- manually records the list of transactions to be submitted
-newtype EmulatedWalletApi a = EmulatedWalletApi { runEmulatedWalletApi :: (ExceptT WalletAPIError (StateT WalletState (Writer [Tx] ))) a }
-    deriving (Functor, Applicative, Monad, MonadState WalletState, MonadWriter [Tx], MonadError WalletAPIError)
+newtype EmulatedWalletApi a = EmulatedWalletApi { runEmulatedWalletApi :: (ExceptT WalletAPIError (StateT WalletState (Writer (WalletLog, [Tx])))) a }
+    deriving (Functor, Applicative, Monad, MonadState WalletState, MonadError WalletAPIError)
+
+-- NB We need to write both [Tx] and WalletLog values, but
```
Comment (michaelpj) [inline wallet-api/src/Wallet/Emulator/Types.hs]: I think I'd rather use the derived instance for the pair and then write `tellTx` and `tellLog` that go to the first or second respectively.

Comment (j-mueller) [reply wallet-api/src/Wallet/Emulator/Types.hs]: but we need `MonadWriter WalletLog` in the WalletAPI actions

Comment (michaelpj) [reply wallet-api/src/Wallet/Emulator/Types.hs]: Ugh, I guess. Or we could add another actual typeclass for that. It's fine.

Comment (j-mueller) [reply wallet-api/src/Wallet/Emulator/Types.hs]: `MonadWriter [Tx]` is only needed to implement `submitTxn`, so I could just add a field to `WalletState` that keeps track of the transactions, and reserve the `MonadWriter` for log messages

File: wallet-api/src/Wallet/API.hs (hunk 1 above)
Comment (j-mueller) [reply wallet-api/src/Wallet/API.hs]: I meant "LoggingWallet is WalletAPI _plus_ the ability  to log messages (...)" but I'll reword it

File: wallet-api/src/Wallet/Emulator/Types.hs (hunk 3 above)
Comment (j-mueller) [reply wallet-api/src/Wallet/Emulator/Types.hs]: > Or we could add another actual typeclass for that

That works, too

Comment (michaelpj) [reply wallet-api/src/Wallet/Emulator/Types.hs]: I don't really mind - I think any of those is fine.

**Comment by michaelpj** [comment]
Nice.

**Pull request merged**